openml/runs/trace.py

```diff
@@ -26,56 +30,348 @@ def get_selected_iteration(self, fold, repeat):
         Returns the trace iteration that was marked as selected. In
         case multiple are marked as selected (should not happen) the
         first of these is returned
-        
+
         Parameters
         ----------
         fold: int
-        
+
         repeat: int
-        
+
         Returns
         ----------
         OpenMLTraceIteration
             The trace iteration from the given fold and repeat that was
             selected as the best iteration by the search procedure
         """
         for (r, f, i) in self.trace_iterations:
-            if r == repeat and f == fold and self.trace_iterations[(r, f, i)].selected is True:
+            if (
+                r == repeat
+                and f == fold
+                and self.trace_iterations[(r, f, i)].selected is True
+            ):
                 return i
-        raise ValueError('Could not find the selected iteration for rep/fold %d/%d' % (repeat, fold))
+        raise ValueError(
+            'Could not find the selected iteration for rep/fold %d/%d' %
+            (repeat, fold)
+        )
+
+    @classmethod
+    def generate(cls, attributes, content):
+        """Generates an OpenMLRunTrace.
+
+        Generates the trace object from the attributes and content extracted
+        while running the underlying flow.
+
+        Returns
+        -------
+        OpenMLRunTrace
+        """
+
+        if content is None:
+            if attributes is None:
+                return None
+            else:
+                raise ValueError('Trace content not available.')
+        if len(content) == 0:
+            raise ValueError('Trace content is empty.')
+        if attributes is not None:
+            if len(attributes) != len(content[0]):
+                raise ValueError(
+                    'Trace_attributes and trace_content not compatible:'
+                    ' %s vs %s' % (attributes, content[0])
+                )
+        else:
+            raise ValueError('No trace attributes available')
+
+        trace = OrderedDict()
+        attribute_idx = {att[0]: idx for idx, att in enumerate(attributes)}
+
+        required_attributes = ['repeat', 'fold', 'iteration', 'evaluation',
```

Comment:
maybe declare these in a more prominent place? (static class variable or global variable. Personally I prefer the static class variable)

Reply:
Although this seems far fetched, we should also check the attributes array whether there are no attributes that carry any of these names (and add a unit test)

Reply:
I added a unit test - however, the parameters will be prefixed with `parameter_`, so this will not be an actual issue.